src/transport_udp.cpp

```diff
@@ -242,361 +520,480 @@ UDPTransport::fd_writer()
+                        a_conn_it->second->report.metrics.total_bytes += rLen;
+                        a_conn_it->second->report.metrics.total_packets++;
+                    }
+
+                    a_conn_it->second->last_rx_msg_tick = current_tick;
+
+                    std::vector<uint8_t> buffer(data + sizeof(hdr), data + sizeof(hdr) + rLen);
+
+                    ConnData cd;
+                    cd.data = buffer;
+                    cd.data_ctx_id = 0; // Currently only implement one data context
+                    cd.conn_id = a_conn_it->second->id;
+
+                    a_conn_it->second->data_contexts[0].rx_data.push(cd);
+
+                    if (a_conn_it->second->data_contexts[0].rx_data.size() < 4) {
+                        delegate.on_recv_notify(cd.conn_id, cd.data_ctx_id);
+                    }
+                }
+                break;
+            }
+            case UdpProtocol::ProtocolType::DATA_DISCARD: {
+                UdpProtocol::DataMsg hdr;
+                memcpy(&hdr, data, sizeof(hdr));
+                rLen -= sizeof(hdr);
+
+                if (a_conn_it != addr_conn_contexts.end()) {
+                    a_conn_it->second->last_rx_msg_tick = _tick_service->get_ticks(std::chrono::milliseconds(1));
+
+                    if (hdr.report_id == 0 || a_conn_it->second->report.report_id != hdr.report_id) {
+
+                        a_conn_it->second->report.report_id = hdr.report_id;
+                        a_conn_it->second->report.metrics.duration_ms = 0;
+                        a_conn_it->second->report.metrics.total_bytes = 0;
+                        a_conn_it->second->report.metrics.total_packets = 0;
+
+                    } else {
+                        a_conn_it->second->report.metrics.duration_ms += current_tick - a_conn_it->second->last_rx_msg_tick;
+                        a_conn_it->second->report.metrics.total_bytes += rLen;
+                        a_conn_it->second->report.metrics.total_packets++;
+                    }
+
+                    a_conn_it->second->last_rx_msg_tick = current_tick;
+                }
+                break;
+            }
+            default:
+                // TODO: Add metric to track discard due to invalid type
+                break;
+        }
     }
 
-    if (rLen == 0) {
-      continue;
+    logger->Log("Done transport reader thread");
+}
+
+TransportError UDPTransport::enqueue(const TransportConnId &conn_id,
+                      const DataContextId &data_ctx_id,
+                      std::vector<uint8_t> &&bytes,
+                      const uint8_t priority,
+                      const uint32_t ttl_ms,
+                      [[maybe_unused]] const EnqueueFlags flags) {
+    if (bytes.empty()) {
+        return TransportError::None;
     }
 
-    std::vector<uint8_t> buffer (data, data + rLen);
 
-    connData cd;
-    cd.data = buffer;
-    cd.streamId = 0;
+    const auto conn_it = conn_contexts.find(conn_id);
```

Comment:
Is a lock needed to protect this map?

Reply:
Yeah, it would be good to add a lock here too.  Updated.